### PR TITLE
Extension-based modularity

### PR DESCRIPTION
**Pull request by dmlloyd**
Introduce two new extensions:

* `quarkus-modular` - this extension contains steps and utilities necessary to assemble a modularized application based on a modularity app model; the extension produces nothing by itself
* `quarkus-jlink` - this extension produces a simple modularized `jlink`-ed application based on the modular model.

When the `quarkus-jlink` extension is present, the default JAR packaging is disabled by default.

Not yet supported:

* Container builds with `jlink`
* Cross-architecture builds with `jlink`
* AOT with `jlink`
* Modular testing

See #51583.

### CODE REVIEW

**Comment by cescoffier** [comment]
🤖 Automatically associated with the **WG - Modularity (JPMS) and JLink** Working Group based on AI classification.

**Comment by cescoffier** [comment]
🤖 Automatically associated with the **WG - AOT Support** Working Group based on AI classification.

**Comment by cescoffier** [comment]
🤖 Automatically associated with the **WG - Quarkus 4** Working Group based on AI classification.

**Comment by dmlloyd** [comment]
Resolved conflicts.

**Comment by dmlloyd** [comment]
Rebased due to conflict, added some code documentation, and added a new log message when an artifact does not have an automatic module name.

**Comment by dmlloyd** [comment]
Added documentation and rebased (again) for conflicts.

**Comment by dmlloyd** [comment]
Should I just keep rebasing this? Seems like a waste of CI budget to me.

File: extensions/packaging/modular/spi/src/main/java/io/quarkus/modular/spi/JarEntryOutputStream.java [hunk 1]
```diff
@@ -0,0 +1,150 @@
...
+            // this is because ZipOutputStream does not support setting general flag bit 3 for STORED entries;
+            // this flag indicates that CRC and size info go into the data descriptor at the end instead.
+            // this is due to PKZIP 2.04g for DOS only supporting this for STORED. that's from 1993...
+            // but ZipInputStream also does not support it now, so it is what it is.
+            if (estSize >= 0 && estSize < 8192) {
+                tempFile = null;
+                raf = null;
+                os = new ByteArrayOutputStream((int) estSize);
+            } else {
+                tempFile = File.createTempFile("quarkus-", "");
```
[the remark this answers is not in the document]
Comment (dmlloyd) [reply extensions/packaging/modular/spi/src/main/java/io/quarkus/modular/spi/JarEntryOutputStream.java]: This is probably legitimate. I'm working on a mitigation.

**Comment by dmlloyd** [comment]
Rebased again for conflicts.

File: bom/application/pom.xml [hunk 2]
```diff
@@ -48,6 +48,7 @@
         <smallrye-common.version>2.18.1</smallrye-common.version>
         <smallrye-config.version>3.17.2</smallrye-config.version>
         <smallrye-health.version>4.3.0</smallrye-health.version>
+        <smallrye-modules.version>1.0.alpha1</smallrye-modules.version>
```
Comment (gsmet) [inline bom/application/pom.xml]: We have software gating solutions that will probably flag this.

is it that alpha?

File: docs/src/main/asciidoc/jlink.adoc [hunk 3]
```diff
@@ -0,0 +1,103 @@
...
+== Prerequisites
+
+The `jlink` extension requires Java 25 or later.
+Earlier JDK versions have compatibility issues with the `jlink` tool that prevent reliable image generation.
+
+== Usage
+
+Add the `quarkus-jlink` extension to your project.
+When the extension is present, a `jlink` image is produced automatically as part of the normal Quarkus build.
+JAR packaging is disabled when `jlink` is active.
```
Comment (gsmet) [inline docs/src/main/asciidoc/jlink.adoc:31]: That's one of the questions I had with this extension-based packaging. Will it be easy to switch to a different packaging if you want to produce several of them?

File: docs/src/main/asciidoc/jlink.adoc [hunk 4]
```diff
@@ -0,0 +1,103 @@
...
+
+`quarkus.jlink.output-directory`::
+The base output directory for the jlink build.
+Type: `Path`.
+Default: derived from `quarkus.package.output-directory`.
+
+`quarkus.jlink.staging-directory`::
+The path of the staging directory used during the jlink build process, relative to the output directory.
+Type: `Path`.
+Default: `staging`.
```
Comment (gsmet) [inline docs/src/main/asciidoc/jlink.adoc:96]: This should use our config doc generation so that it's updated automatically. I'll do it once this is in.

File: extensions/packaging/jlink/deployment/src/main/java/io/quarkus/jlink/deployment/JLinkConfig.java [hunk 5]
```diff
@@ -0,0 +1,62 @@
...
+
+/**
+ * Configuration for producing a {@code jlink} image.
+ */
+@ConfigRoot
+@ConfigMapping(prefix = "quarkus.jlink")
+// @WithDynamicDefaults(JLinkConfigDefaults.class)
+public interface JLinkConfig {
+    /**
+     * {@return {@code true} to enable {@code jlink} execution, or {@code false} otherwise}
```
Comment (gsmet) [inline extensions/packaging/jlink/deployment/src/main/java/io/quarkus/jlink/deployment/JLinkConfig.java:19]: I'm not sure how this will render in the config doc. We usually, just have regular Javadoc for these methods (mostly because they were initially fields).

I'll check when I switch to using the generated config doc.

File: extensions/packaging/jlink/runtime/pom.xml [hunk 6]
```diff
@@ -0,0 +1,77 @@
...
+    <parent>
+        <groupId>io.quarkus</groupId>
+        <artifactId>quarkus-jlink-parent</artifactId>
+        <version>999-SNAPSHOT</version>
+    </parent>
+
+    <artifactId>quarkus-jlink</artifactId>
+
+    <name>Quarkus - JLink - Runtime</name>
+
```
Comment (gsmet) [inline extensions/packaging/jlink/runtime/pom.xml:17]: I think we need a description here for code.quarkus.io.

File: extensions/packaging/modular/runtime/src/main/resources/META-INF/quarkus-extension.yaml [hunk 7]
```diff
@@ -0,0 +1,9 @@
+---
+artifact: ${project.groupId}:${project.artifactId}:${project.version}
+name: "Modularity"
+metadata:
+  categories:
+  - "packaging"
+  status: "experimental"
+  config:
+  - "quarkus.modular."
```
Comment (gsmet) [inline extensions/packaging/modular/runtime/src/main/resources/META-INF/quarkus-extension.yaml:9]: I wonder if we should make this one unlisted as I would expect it's not something the user consume directly?

File: extensions/packaging/modular/spi/src/main/java/io/quarkus/modular/spi/items/AddedDependencyBuildItem.java [hunk 8]
```diff
@@ -0,0 +1,50 @@
...
+
+import io.quarkus.builder.item.MultiBuildItem;
+import io.smallrye.common.constraint.Assert;
+import io.smallrye.modules.desc.Dependency;
+import io.smallrye.modules.desc.Modifiers;
+
+/**
+ * A build item which adds an extra dependency between modules.
+ */
+public final class AddedDependencyBuildItem extends MultiBuildItem {
```
Comment (gsmet) [inline extensions/packaging/modular/spi/src/main/java/io/quarkus/modular/spi/items/AddedDependencyBuildItem.java]: Should we go with `AddModuleDependencyBuildItem`? @aloubyansky ended up being confused about it so maybe we should be explicit?

Also, for me `Added` looks like something you would consume to inspect the added modules, not to actively add one.

**Review by gsmet — DISMISSED** [review]
I added a couple of very small comments. Nothing blocking really so, if you prefer, they can be addressed as follow-ups if you think they should.

File: bom/application/pom.xml (hunk 2 above)
Comment (dmlloyd) [reply bom/application/pom.xml]: The API can change without notice, which makes it "alpha" to me. I could re-release it as beta if you want though; I don't feel strongly about it.

File: extensions/packaging/modular/spi/src/main/java/io/quarkus/modular/spi/items/AddedDependencyBuildItem.java (hunk 8 above)
Comment (gastaldi) [reply extensions/packaging/modular/spi/src/main/java/io/quarkus/modular/spi/items/AddedDependencyBuildItem.java]: `Additional` is often used in our build items

File: docs/src/main/asciidoc/jlink.adoc (hunk 3 above)
Comment (dmlloyd) [reply docs/src/main/asciidoc/jlink.adoc:31]: Yes, the user should add an extension for desired packaging type. If you want the old packaging too, change the `quarkus.package.jar.enabled` back to `true` again.

File: extensions/packaging/modular/runtime/src/main/resources/META-INF/quarkus-extension.yaml (hunk 7 above)
Comment (dmlloyd) [reply extensions/packaging/modular/runtime/src/main/resources/META-INF/quarkus-extension.yaml:9]: Sure. It is something that extension authors should consume; does that matter?

File: extensions/packaging/modular/spi/src/main/java/io/quarkus/modular/spi/items/AddedDependencyBuildItem.java (hunk 8 above)
Comment (dmlloyd) [reply extensions/packaging/modular/spi/src/main/java/io/quarkus/modular/spi/items/AddedDependencyBuildItem.java]: Sure, I can make that change.

**Comment by dmlloyd** [comment]
Looks like an indirect conflict. Resolving it now.

**Comment by dmlloyd** [comment]
#54321 will fix the issue; once it's merged I'll rebase again...

**Comment by dmlloyd** [comment]
Rebased again. This time updating to `smallrye-modules` version `1.0.beta1`, adding a dependabot thingy, fixing a minor bug around the `synthetic` modifier using a new API, and automatically detecting JDK modules to add to JLink (which is something I meant to do but forgot to finish).

**Comment by dmlloyd** [comment]
Also pushed up a minor change to ensure that the `--add-modules` arguments always list the modules in the same order every time.

File: independent-projects/bootstrap/app-model/src/main/java/io/quarkus/maven/dependency/ResolvedDependency.java [hunk 9]
```diff
@@ -79,4 +93,116 @@ default PathTree getContentTree(PathFilter pathFilter) {
...
+            String moduleName = ma.automaticModuleName();
+            if (moduleName != null) {
+                return moduleName;
+            }
+        }
+        // next, see if we have a (temporary) policy-defined module name for this artifact
+        String groupId = dep.getGroupId();
+        String artifactId = dep.getArtifactId();
+        String classifier = dep.getClassifier();
+        // XXX MANUAL AUTOMATIC MODULE NAME FIXUPS
```
Comment (gastaldi) [inline independent-projects/bootstrap/app-model/src/main/java/io/quarkus/maven/dependency/ResolvedDependency.java:125]: Can we do something to avoid these switch statements?

Comment (dmlloyd) [reply independent-projects/bootstrap/app-model/src/main/java/io/quarkus/maven/dependency/ResolvedDependency.java:125]: Not right now. The ongoing solution will be to open issues & PRs for each project corresponding to one of these names to get their automatic module name set and/or establish an actual descriptor.

Comment (dmlloyd) [reply independent-projects/bootstrap/app-model/src/main/java/io/quarkus/maven/dependency/ResolvedDependency.java:125]: Note that some of these may already be fixed; we can remove them as they get released.

File: extensions/packaging/jlink/deployment/src/main/java/io/quarkus/jlink/deployment/JLinkConfigCustomizer.java [hunk 10]
```diff
@@ -0,0 +1,18 @@
+package io.quarkus.jlink.deployment;
+
+import io.smallrye.config.SmallRyeConfigBuilder;
+import io.smallrye.config.SmallRyeConfigBuilderCustomizer;
+
+/**
+ * The factory for the defaults interceptor.
+ */
+// todo: remove if we get @WithDynamicDefaults
+public final class JLinkConfigCustomizer implements SmallRyeConfigBuilderCustomizer {
```
Comment (gastaldi) [inline extensions/packaging/jlink/deployment/src/main/java/io/quarkus/jlink/deployment/JLinkConfigCustomizer.java:10]: Looks like dead code?

Comment (dmlloyd) [reply extensions/packaging/jlink/deployment/src/main/java/io/quarkus/jlink/deployment/JLinkConfigCustomizer.java:10]: ~~It's loaded as a service.~~ wait, it's commented out for some reason. I'll look into it.

**Comment by gastaldi** [comment]
It would be nice if we had a couple of tests

File: extensions/packaging/jlink/deployment/src/main/java/io/quarkus/jlink/deployment/JLinkConfigCustomizer.java (hunk 10 above)
Comment (gastaldi) [reply extensions/packaging/jlink/deployment/src/main/java/io/quarkus/jlink/deployment/JLinkConfigCustomizer.java:10]: Yeah, I mean: it doesn't do anything? Because the body is commented out

**Comment by dmlloyd** [comment]
Tests would be good. But, I'm not sure how to even approach it, since it doesn't fit into the existing testing infrastructure. I could probably [have Claude] add some unit tests for the code within the extension classes, but anything else would need an integration test using a framework that doesn't yet exist.

File: docs/src/main/asciidoc/jlink.adoc [hunk 11]
```diff
@@ -0,0 +1,103 @@
...
+
+The `jlink` extension produces a custom Java runtime image for a Quarkus application.
+The image contains only the JDK modules and application modules that are needed to run the application, resulting in a smaller, faster, and more self-contained distribution.
+This extension depends on the xref:modularity.adoc[Quarkus modularity extension] to compute the application's module graph.
+
+include::{includes}/extension-status.adoc[]
+
+== Prerequisites
+
+The `jlink` extension requires Java 25 or later.
```
Comment (gastaldi) [inline docs/src/main/asciidoc/jlink.adoc:24]: Is this is a hard requirement or just tested-on?

Comment (gastaldi) [reply docs/src/main/asciidoc/jlink.adoc:24]: Ah, the paragraph below explains it, nvm

Comment (dmlloyd) [reply docs/src/main/asciidoc/jlink.adoc:24]: Somewhere in between. The extension "works" on 24 (and maybe earlier) but `jlink` itself is broken/slightly incompatible on those versions.

**Comment by gastaldi** [comment]
I asked Claude to review and this is the output it produced: 

## Code Review: PR #53140 — Extension-based modularity

### Overview

This PR introduces JPMS (Java Platform Module System) support to Quarkus through two new experimental extensions:

- **`quarkus-modular`** — Builds a complete module graph for the application at build time, enhancing JPMS `module-info.class` descriptors and computing descriptors for automatic modules from Maven dependency graphs.
- **`quarkus-jlink`** — Produces a custom JDK runtime image via `jlink`, containing only the modules the application needs.

The architecture splits modules into "boot modules" (loaded by the JDK boot layer) and "dynamic modules" (loaded at runtime by `smallrye-modules`). The PR adds ~4500 lines across 54 files, organized in 11 incremental commits.

---

### Strengths

- **Solid architecture.** The boot/dynamic module split is well-designed. Using `smallrye-modules` for dynamic loading avoids fighting JPMS's inflexible built-in layer system while still getting strong encapsulation benefits.
- **Good use of Quarkus build items.** The SPI (`BootModulePathBuildItem`, `AddDependencyBuildItem`, `ModuleOpenBuildItem`, etc.) follows established Quarkus patterns and will let individual extensions opt into modularity incrementally.
- **Security-conscious.** Patch 05 explicitly restricts temp file permissions using POSIX/ACLs — good catch for `JarEntryOutputStream`.
- **Well-documented.** Three new `.adoc` files cover both user-facing usage and extension-author SPI. The modularity-reference doc is especially useful.
- **Incremental commits.** The 11-patch structure is clean and reviewable: base code, docs, comments, rename, API upgrade, each in its own commit.

---

### Issues and Suggestions

#### Critical: No Tests

The PR adds ~4500 lines of new code with **zero test files**. For experimental extensions this might be acceptable as a first merge, but it should be called out:

- `ModularitySteps` is ~960 lines of complex graph computation with hardcoded fixups — this is the most fragile part and most needs testing.
- `JarEntryOutputStream` has non-trivial STORED/DEFLATED logic with temp files and CRC32 precomputation — edge cases here could produce corrupt JARs.
- `ModuleWriter` has XML and `module-info.class` generation — output correctness can only be validated with tests.
- `ResolvedDependency.computeModuleName()` has a manual fixup table and a multi-step heuristic — unit tests would document expected behavior.

**Recommendation:** At minimum, add unit tests for `computeModuleName()`, `JarEntryOutputStream`, and `ModuleWriter`. Integration tests for the full `ModularitySteps` pipeline should follow before removing the experimental label.

#### High: Hardcoded Module Fixups in `ModularitySteps`

The `switch (moduleName)` block (~lines 715-785 of `ModularitySteps.java`) contains hardcoded fixups for Vert.x, RESTEasy, OIDC, Hibernate, Jackson, and others. While each has a `todo` comment, this creates a maintenance burden:

```java
switch (moduleName) {
    case "io.quarkus.vertx" -> { ... }
    case "org.jboss.resteasy.core.spi" -> { ... }
    case "io.quarkus.oidc.common" -> { ... }
    // ~15 more cases
}
```

The todos reference upstream releases that would fix these. **Suggestion:** Track these in a single issue with a checklist so they don't become permanent fixtures. Consider a data-driven approach (e.g., a resource file mapping module names to fixups) so they're easier to audit.

#### High: Hardcoded Module Name Fixup Table in `ResolvedDependency`

`computeModuleName()` contains a large `switch` with ~30 manual overrides:

```java
case "io.quarkus", "quarkus-arc" -> "io.quarkus.arc.runtime";
case "io.smallrye", "smallrye-context-propagation" -> "io.smallrye.context";
// etc.
```

This is brittle — any new artifact without a proper `Automatic-Module-Name` manifest entry or `module-info.class` will need to be added here. **Suggestion:** File a tracking issue for upstream projects to add `Automatic-Module-Name` to their manifests, and log warnings (which patch 03 does) to surface gaps.

#### Medium: Alpha/Beta Dependency

`smallrye-modules` was updated from `1.0.alpha1` to `1.0.beta1` in patch 09. The API churn between patches (e.g., `Modifiers<Modifier>` → `Modifier.Set`, `Modifier.set()` → `Modifier.Set.of()`) confirms the library is still stabilizing. This is acceptable for an experimental extension but should be noted in the docs.

#### Medium: Commented-Out Code

`JLinkConfigCustomizer.java` contains:

```java
// @Override
// public void configCustomizer(SmallRyeConfigBuilder builder) {
//     builder.withInterceptors(new JLinkConfigDefaultsInterceptor());
// }
```

This should either be enabled or removed. If it's needed for a future change, a comment explaining when to uncomment would be better than dead code.

#### Medium: `sneak()` Pattern

`ModuleWriter.java` uses an unchecked exception bypass:

```java
@SuppressWarnings("unchecked")
private static <E extends Throwable> void sneak(Throwable t) throws E {
    throw (E) t;
}
```

This circumvents Java's checked exception system. While common in some codebases, it can confuse callers who don't expect `IOException` from methods that don't declare it. Consider using `io.smallrye.common.function.ExceptionFunction` or similar wrappers from libraries already in the dependency tree.

#### Medium: Silent Exception Handling

In `ModularitySteps`, service file reading silently catches and ignores `IOException`:

```java
} catch (IOException ignored) {
    // ignore
}
```

At minimum, log at DEBUG/TRACE level so issues are diagnosable.

#### Low: `JLinkConfig` Default Launcher Name

```java
@WithDefault("my-app")  // TODO: use quarkus.package.output-name
String launcherName();
```

The TODO is tracked but "my-app" will surprise users. This should be resolved before the extension leaves experimental status.

#### Low: `jdkModulesUsed` Changed from `Set` to `List`

Patch 11 changes `jdkModulesUsed` from `Set<String>` to `List<String>` with manual binary-search insertion to maintain sort order. This looks like it's optimizing for deterministic `--add-modules` output. A `TreeSet` would be simpler and equally sorted:

```java
// Current: manual binary search on ArrayList
int idx = Collections.binarySearch(jdkModulesUsed, moduleName);
if (idx < 0) {
    jdkModulesUsed.add(-idx - 1, moduleName);
}

// Simpler:
jdkModulesUsed = new TreeSet<>();
jdkModulesUsed.add(moduleName);
```

#### Low: Windows ACL Entry Missing Permissions

In `JarEntryOutputStream`, the Windows ACL entry grants `CREATOR OWNER` an `ALLOW` entry but specifies **no permissions** (no `AclEntryPermission` values). This might result in an empty ACL that denies all access on some Windows versions. Verify this works correctly on Windows, or add explicit `READ_DATA`/`WRITE_DATA` permissions:

```java
AclEntry.newBuilder()
    .setType(AclEntryType.ALLOW)
    .setPrincipal(...)
    .setPermissions(AclEntryPermission.READ_DATA, AclEntryPermission.WRITE_DATA)
    .build()
```

#### Low: `ibm.` Module Prefix

Patch 11 adds `n.startsWith("ibm.")` to the JDK module detection. This is good for IBM Semeru/OpenJ9 support but should be documented somewhere (e.g., a comment explaining which JDK distributions provide `ibm.*` modules).

#### Low: Duplicate JDK Module Tabulation Code

The JDK module tabulation logic appears twice — once for regular modules and once for the app module:

```java
mi.dependencies().stream()
    .map(DependencyInfo::moduleName)
    .filter(n -> n.startsWith("java.") || n.startsWith("jdk.") || n.startsWith("ibm."))
    .forEach(usedJdkModuleNames::add);
```

Extract this into a private method to avoid the duplication.

---

### Documentation

The three new `.adoc` files are well-structured and follow Quarkus conventions. Minor suggestions:

- `jlink.adoc` says "Java 25 or later" — clarify whether this is a hard requirement or just tested-on.
- The config reference in `jlink.adoc` manually lists properties. Consider using `@ConfigMapping`-generated config reference once the extension stabilizes.
- `modularity.adoc` could benefit from a simple diagram showing the boot/dynamic module split.

---

### Summary

This is a well-architected PR that introduces a major new capability. The main concerns are:

1. **No tests** — the number 1 issue: Complex graph algorithms and JAR manipulation need test coverage.
2. **Extensive hardcoded fixups** — workable for experimental status but need a tracking plan.
3. **Pre-release dependency** — `smallrye-modules 1.0.beta1` is still evolving.

The code quality is high, the incremental commit structure is clean, and the documentation is thorough. Once test coverage is added and the fixup tables have tracking issues, this will be a solid foundation for JPMS support in Quarkus.

**Review by gastaldi — DISMISSED** [review]
Let's do it !

**Review by gastaldi — DISMISSED** [review]

**Comment by gastaldi** [comment]
Can you please squash everything so we can approve and merge?

**Comment by dmlloyd** [comment]
Are you sure you want me to squash it? I'm not going to un-squash once I've squashed.

**Comment by gastaldi** [comment]
I'd prefer to, because most commits are fixups that don't really add up to the main tree. But I'll approve it anyway

**Review by gastaldi — APPROVED** [review]

**Comment by dmlloyd** [comment]
OK, one more push with the squashed commit.

**Review by gastaldi — DISMISSED** [review]

**Comment by dmlloyd** [comment]
Good catch. I didn't notice any failure on Windows.

**Comment by dmlloyd** [comment]
Just leave this as is and I'll look into the windows failures on tuesday.

**Comment by dmlloyd** [comment]
I feel pretty confident about the Windows failure now (as confident as I can be without actually testing it directly myself anyway). If Windows passes, let's ship it.

**Review by gastaldi — APPROVED** [review]

**Comment by gastaldi** [comment]
Looks like the Windows tests are still failing

**Review by gastaldi — CHANGES_REQUESTED** [review]
Windows tests are still failing

**Comment by dmlloyd** [comment]
Ah. Claude found the problem in 2 seconds (literally). Fix incoming.

**Comment by dmlloyd** [comment]
Fix pushed; also rebased for good measure.

**Comment by gastaldi** [comment]
~~What was the fix BTW? Just out of curiosity :)~~

nvm, found it: 
```java
                        .setPermissions(EnumSet.allOf(AclEntryPermission.class))
```

**Review by gastaldi — APPROVED** [review]

**Comment by gastaldi** [comment]
Why are these utility classes (JarEntryOutputStream, LogWriter, etc) included here? Are they supposed to be part of smallrye-common at some point or?

**Comment by dmlloyd** [comment]
Yeah I'll probably move them over eventually. I have a branch for JAR/ZIP support in `smallrye-common` that would probably fit this. It's non-API so moving it later shouldn't be a problem.

**Comment by cescoffier** [comment]
I just want to say "WOOHOO!"
